internal/collect/tags.go

```diff
@@ -0,0 +1,118 @@
+package collect
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/vmware/govmomi/vapi/tags"
+
+	"github.com/newrelic/nri-vsphere/internal/load"
+	"github.com/vmware/govmomi/vim25/mo"
+	"github.com/vmware/govmomi/vim25/types"
+)
+
+// separate in chunks of 2000 objects following performance recommendation
+// https://www.vmware.com/content/dam/digitalmarketing/vmware/en/pdf/techpaper/performance/tagging-vsphere67-perf.pdf
+const maxBatchSize = 2000
+
+// TagCategories retreive all tag categories from vcenter and store them for future use
+func collectTagCategories(tc load.TagCategories, tm *tags.Manager) error {
+	ctx := context.Background()
+	categories, err := tm.GetCategories(ctx)
+	if err != nil {
+		return err
+	}
+	for _, c := range categories {
+		tc[c.ID] = c.Name
+	}
+	return nil
+}
+
+func collectTags(config *load.Config, managedObjectsSlice interface{}, dc *load.Datacenter) error {
+	var ref []mo.Reference
+
+	if !(config.Args.EnableVsphereTags && config.IsVcenterAPIType) {
+		return nil
+	}
+
+	switch obs := managedObjectsSlice.(type) {
```

Comment:
why do we need this swich case? 
Is it possible to have the type outside this set?

Reply:
i done it this way to simply call this func from every collector passing the slice of collected objects. I didn't find a way to do this better without generics

Reply:
mmm I was thinking to ask for a less generic interface, something like:
```
func collectTags(config *load.Config, obs []mo.Entity, dc *load.Datacenter) error {
	var ref []mo.Reference

	if !(config.Args.EnableVsphereTags && config.IsVcenterAPIType) {
		return nil
	}

	for _, o := range obs {
		ref = append(ref, o.Reference())
	}
```

The issue is that the function ask for a slice of interface and there is not a way to cast it without using reflection 😢 
> https://stackoverflow.com/questions/12753805/type-converting-slices-of-interfaces